### PR TITLE
Add multiD Gemm client APIs

### PR DESCRIPTION
**Pull request by ltqin**

### CODE REVIEW

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_multiple_d_dl.hpp [hunk 1]
```diff
@@ -0,0 +1,669 @@
+// SPDX-License-Identifier: MIT
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/impl/device_gemm_multiple_d_dl.hpp:1]: Also a dlops version, cool!

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp [hunk 2]
```diff
@@ -172,6 +172,35 @@ struct AddAdd
     }
 };
 
+// C = A * B
+// E = (C + D0) x D1
+struct AddMultiply
+{
+    template <typename E, typename C, typename D0, typename D1>
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/element/element_wise_operation.hpp:179]: Please do like this if only certain datatype are supported

image

**Review by asroy — CHANGES_REQUESTED** [review]
a minor change needed

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp (hunk 2 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/element/element_wise_operation.hpp:179]: done

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp [hunk 3]
```diff
@@ -172,6 +172,42 @@ struct AddAdd
...
+    __host__ __device__ void operator()(E& e, const C& c, const D0& d0, const D1& d1) const;
+
+    template <>
+    __host__ __device__ void operator()<half_t, half_t, half_t, half_t>(half_t& e,
+                                                                        const half_t& c,
+                                                                        const half_t& d0,
+                                                                        const half_t& d1) const
+    {
+        const half_t y = c + d0 + d1;
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/element/element_wise_operation.hpp:188]: Shouldn't you multiply by `d1`  as in operation comment and name? 
```suggestion
        const half_t y = (c + d0) * d1;
```

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp [hunk 4]
```diff
@@ -172,6 +172,42 @@ struct AddAdd
...
+        e              = y;
+    }
+    template <>
+    __host__ __device__ void operator()<half_t, float, half_t, half_t>(half_t& e,
+                                                                       const float& c,
+                                                                       const half_t& d0,
+                                                                       const half_t& d1) const
+    {
+        const half_t y = type_convert<half_t>(c) + d0 + d1;
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/element/element_wise_operation.hpp:197]: Shouldn't you multiply by `d1` as in operation comment and name?
```suggestion
        const half_t y = (type_convert<half_t>(c) + d0) * d1;
```

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp [hunk 5]
```diff
@@ -172,6 +172,42 @@ struct AddAdd
...
+        e              = y;
+    }
+    template <>
+    __host__ __device__ void operator()<float, float, half_t, half_t>(float& e,
+                                                                      const float& c,
+                                                                      const half_t& d0,
+                                                                      const half_t& d1) const
+    {
+        const half_t y = type_convert<half_t>(c) + d0 + d1;
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/element/element_wise_operation.hpp:206]: Shouldn't you multiply by `d1` as in operation comment and name?
```suggestion
        const half_t y = (type_convert<half_t>(c) + d0) * d1;
```

File: library/include/ck/library/tensor_operation_instance/gpu/gemm_add_multiply.hpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                                                    Row,
+                                                    F16,
+                                                    F16,
+                                                    F16_F16_Tuple,
+                                                    F16,
+                                                    PassThrough,
+                                                    PassThrough,
+                                                    AddMultiply>>>&);
+
+// GEMM + Add + Add + FastGelu
```
Comment (aosewski) [inline library/include/ck/library/tensor_operation_instance/gpu/gemm_add_multiply.hpp:74]: Is this comment correct?

File: profiler/include/profiler/profile_gemm_add_multiply_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+          typename AccDataType,
+          typename D0DataType,
+          typename D1DataType,
+          typename EDataType,
+          typename ALayout,
+          typename BLayout,
+          typename D0Layout,
+          typename D1Layout,
+          typename ELayout>
+bool profile_gemm_add_add_fastgelu_impl(int do_verification,
```
Comment (aosewski) [inline profiler/include/profiler/profile_gemm_add_multiply_impl.hpp:36]: Is this function name correct?

**Review by aosewski — CHANGES_REQUESTED** [review]
Seems to me that elementwise operation has incorrect implementation.

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp (hunk 4 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/element/element_wise_operation.hpp]: done

File: include/ck/tensor_operation/gpu/element/element_wise_operation.hpp (hunk 5 above)
Comment (ltqin) [reply include/ck/tensor_operation/gpu/element/element_wise_operation.hpp]: done

File: library/include/ck/library/tensor_operation_instance/gpu/gemm_add_multiply.hpp (hunk 6 above)
Comment (ltqin) [reply library/include/ck/library/tensor_operation_instance/gpu/gemm_add_multiply.hpp]: done

File: profiler/include/profiler/profile_gemm_add_multiply_impl.hpp (hunk 7 above)
Comment (ltqin) [reply profiler/include/profiler/profile_gemm_add_multiply_impl.hpp]: done

**Review by aosewski — APPROVED** [review]

**Review by asroy — APPROVED** [review]

**Pull request merged**